src/libmongoc/src/mongoc/mongoc-client.c

```diff
@@ -957,13 +957,8 @@ _mongoc_client_recv (mongoc_client_t *client,
    BSON_ASSERT (buffer);
    BSON_ASSERT (server_stream);
 
-   if (!mongoc_cluster_try_recv (
-          &client->cluster, rpc, buffer, server_stream, error)) {
-      mongoc_topology_invalidate_server (
-         client->topology, server_stream->sd->id, error);
-      return false;
-   }
-   return true;
+   return mongoc_cluster_try_recv (
```

Comment:
`mongoc_topology_invalidate_server` marks a server description as Unknown.
I don't think it should be called in isolation, since there is other behavior that needs to happen when marking a server Unknown due to a network error (generation check, clear connection pool, and cancelling the streamable ismaster). Invalidating a server on a network error without the other logic would be a bug.

Fortunately `mongoc_cluster_recv` already does error handling on a network error... except for failure to decompress. But that behavior is already inconsistent (https://jira.mongodb.org/browse/CDRIVER-3694) and I'm also not positive a decompression failure is a network error.

I added at least a basic test to ensure that a hangup on `_mongoc_client_recv` does the right error handline.